Jenkinsfile

```diff
@@ -20,6 +20,25 @@ pipeline {
       }
     }
 
+    stage ('Clean') {
+      steps {
+        container('builder') {
+          script { sh 'make clean' }
+        }
+      }
+    }
+
+    stage ('Building Gems') {
+      steps {
+        container ('builder') {
+          sh 'make base.gem'
+          sh 'make api.gem'
+          sh 'make web.gem'
+          sh 'make engine.gem'
+        }
+      }
+    }
```

Comment:
Building gems => only if we are building a tag. We don't need it on master and/or PR.

Reply:
And I would build and push the gems (as part of the same stage) after the tests successfully ran and the images successfully built.

That is, the Jenkinsfile truly must have an invariant saying "ONLY IF everything seems 100% fine, then we actually release images & gems (since that has a external effect on other systems)".